### PR TITLE
Fix total (plasma + kinetic neutral) deuterium particle balance: wall recombination (corr_neg and bcs(:) type)

### PR DESCRIPTION
**Pull request by Daniel-Maris**
This PR fixes the total deuterium particle balance (plasma + kinetic neutrals) by using the same correction on the particle side as on the fluid side for negative T and ne (corr_neg rather than previously used max(ne,1e16)), and by taking into account which boundary condition is locally on the fluid side through the bcs(:).

This breaks the reg test so those are updated.

This was already a PR before github, so for details see https://git.iter.org/projects/STAB/repos/jorek/pull-requests/1031/overview and https://jira.iter.org/browse/IMAS-5945 if you have access.

### CODE REVIEW

File: diagnostics/mod_integrals3D.f90 [hunk 1]
```diff
@@ -921,7 +921,7 @@ subroutine int3d_new(my_id, node_list, element_list, bnd_node_list, bnd_elm_list
                       !- (gamma-1.d0)* aux_mom_par0 * vpar0 * BigR *xjac* delta_phi *wst
 #endif
           !>Lost to recombination (no Brehmstralung)
-          local_Prec = local_Prec + r0_corr*r0_corr*(T0_corr*Srec_T_ncs)*BigR *xjac* delta_phi *wst
+          local_Prec = local_Prec + r0_corr*r0_corr*(Te0_corr*Srec_T_ncs)*BigR *xjac* delta_phi *wst
```
Comment (N-Schwarz) [inline diagnostics/mod_integrals3D.f90]: I think this should be the total temperature. Can you double check?

Comment (Daniel-Maris) [reply diagnostics/mod_integrals3D.f90]: Hm, good catch, now I don't know to be honest. 
@szmate00 could you check this?
I find these names all a bit confusing. I didn't expect Bremsstrahlung to be an important radiation mechanism for recombination, I thought it was mostly line radiation from the electron dropping down into a lower state. Is the term abused here, or where does it come from? Is it the 3rd electron in the 3 body recombination? 

There are the ions and electrons getting neutral, so their energy leaving the plasma side. Now if that was all we cared about we could just use T0 here and be done, but to be consistent with the neutrals side, I think what is instead done is that the ion loss should be calculated here (so actually Ti0 should be used), while the electron losses are calculated separately in the line below (Prb). This is my understanding because there are equivalent terms in elt_matrix_fft (lines 1865, 1914, 2051 and 2054) of model600.

So @szmate00, it should be Ti0 here, right?

Comment (N-Schwarz) [reply diagnostics/mod_integrals3D.f90]: You're right it should be Ti to be consistent with the element matrix

Comment (szmate00) [reply diagnostics/mod_integrals3D.f90]: Yes, Ti0, I think

Comment (N-Schwarz) [reply diagnostics/mod_integrals3D.f90]: Thank you,m can you correct it please?

Comment (Daniel-Maris) [reply diagnostics/mod_integrals3D.f90]: Yes, done. Thanks for the reminder

File: particles/mod_fields.f90 [hunk 2]
```diff
@@ -313,21 +314,18 @@ subroutine calc_NeTeTi(fields,time,i_elm,st,phi,                   &
...
+  if (present(n_e_raw)) n_e_raw = P(1) * n_norm
+  n_e = corr_neg_dens(P(1)) * n_norm
 
   ! temperatures
   ! compute Te (required)
-  T_e_temp = P(ii_Te) * T_norm
-  if (present(T_e_raw)) T_e_raw = T_e_temp
-  T_e = max(T_e_temp, T_FLOOR)
+  if (present(T_e_raw)) T_e_raw = P(ii_Te) * T_norm
+  T_e = corr_neg_temp(P(ii_Te)) * T_norm
```
Comment (N-Schwarz) [inline particles/mod_fields.f90:323]: I think to be consistent, we should do the correction on the total temperature. Same below

File: particles/mod_fields.f90 [hunk 3]
```diff
@@ -438,12 +417,12 @@ subroutine calc_NjTj(fields, time, i_elm, st, phi, m_i_over_m_imp, ne, te, ni, t
   if(with_TiTe) then
      if(with_impurities) then
         call fields%interp_PRZ(time,i_elm,[var_rho,var_Te,var_rhoimp,var_Ti],4,st(1),st(2),phi,P,P_s,P_t,P_phi,P_time,R,R_s,R_t,Z,Z_s,Z_t)
-        Ti = max(P(4)/(K_BOLTZ*MU_ZERO*central_density*1.d20), 1.d0)
+        Ti = corr_neg_temp(P(4))/(K_BOLTZ*MU_ZERO*central_density*1.d20)
```
Comment (N-Schwarz) [inline particles/mod_fields.f90:420]: Same here, the correction should be on the total temperature

**Review by N-Schwarz — COMMENTED** [review]
Hi @Daniel-Maris , thanks. This looks good, just if you want to be consistent I think it would be better to apply the correction to the total temperature everywhere and also change it in mod_boundary_matrix_open.f90.

File: particles/mod_fields.f90 (hunk 2 above)
Comment (Daniel-Maris) [reply particles/mod_fields.f90:323]: Yes, we discussed this before the move to github on the old PR page; I had already committed this but since it changes the underlying fluid mhd model, breaking many regtests in the process, we thought it would be better to do this in a separate PR to make it more clear what this PR is exactly doing and what that PR is exactly doing. If you now prefer to include this in this PR anyway please let me know. I'm planning to do that PR once this is finished to avoid conflicts with this PR

Comment (N-Schwarz) [reply particles/mod_fields.f90:323]: Sorry, I thought you meant to fix this already here.
We can do this in a later one.

**Comment by N-Schwarz** [comment]
@quietF , @szmate00 Can you have a look at this pull request?

**Comment by N-Schwarz** [comment]
You are probably very busy,  @szmate00 , @quietF. This is a gentle reminder, please have a look when you find the time

**Review by szmate00 — APPROVED** [review]

**Comment by szmate00** [comment]
Approved! Thank you Daniel, really good work!

File: diagnostics/mod_integrals3D.f90 [hunk 4]
```diff
@@ -921,7 +921,7 @@ subroutine int3d_new(my_id, node_list, element_list, bnd_node_list, bnd_elm_list
                       !- (gamma-1.d0)* aux_mom_par0 * vpar0 * BigR *xjac* delta_phi *wst
 #endif
           !>Lost to recombination (no Brehmstralung)
-          local_Prec = local_Prec + r0_corr*r0_corr*(T0_corr*Srec_T_ncs)*BigR *xjac* delta_phi *wst
+          local_Prec = local_Prec + r0_corr*r0_corr*(Ti0_corr*Srec_T_ncs)*BigR *xjac* delta_phi *wst
```
Comment (quietF) [inline diagnostics/mod_integrals3D.f90:924]: Hi @Daniel-Maris  
I am thinking this should be Te0_corr

File: particles/mod_particle_wall_interaction.f90 [hunk 5]
```diff
@@ -1733,42 +1744,81 @@ subroutine project_sputter_vars_on_edge(this, sim)
...
+    !$omp private(i, n_e, n_e_corr, T_e, vpar, grad_n_e, E, B, psi, U, psi_norm, vector_normal, B_hat, cos_alpha, grad_par_n_e, q, T_i, mass_ion, c_s, Gamma_d, &
+    !$omp         Gamma_convective, yield, inodes, inode, node, n_dirichlet_BC_nodes, n_sheath_BC_nodes, n_grad0_BC_nodes, &
+    !$omp         D_par_si, D_perp_si, ierr) schedule(static)
     do i = 1, size(this%fluid_yield_integral%patch(i_patch)%xyz, 2) !< over all nodes
-      call sim%fields%calc_NeTevpar(sim%time, this%fluid_yield_integral%patch(i_patch)%i_elm_jorek_edge(i), this%fluid_yield_integral%patch(i_patch)%st(:,i), &
-        real(this%fluid_yield_integral%patch(i_patch)%xyz(3,i), 8), n_e, T_e, vpar)
+      call sim%fields%calc_NeTeTi(sim%time, this%fluid_yield_integral%patch(i_patch)%i_elm_jorek_edge(i), this%fluid_yield_integral%patch(i_patch)%st(:,i), &
+        real(this%fluid_yield_integral%patch(i_patch)%xyz(3,i), 8), n_e_corr, T_e, T_i, n_e_raw=n_e, grad_raw_n_e=grad_n_e)
+
+      call sim%fields%calc_vpar(sim%time, this%fluid_yield_integral%patch(i_patch)%i_elm_jorek_edge(i), this%fluid_yield_integral%patch(i_patch)%st(:,i), &
```
Comment (quietF) [inline particles/mod_particle_wall_interaction.f90:1756]: @Daniel-Maris 
have you checked whether separately calculating vpar and neTeTi leads to a change in performance? I would have guessed that asking for the interpolation of all 3 (or 4 if Ti is used) is faster than 2 + 1.
